document-generator-prosecution/src/main/java/uk/gov/companieshouse/document/generator/prosecution/handler/ProsecutionHandler.java

```diff
@@ -0,0 +1,125 @@
+
+    /**
+     * Method to get the base prosecution information for dealing with prosecution documents
+     * Throws a HandlerException if any exceptions occurred in the service
+     * 
+     * @param resourceUri the URI for the defendant
+     * @return the ProsecutionDocument containing defendant, prosecution case and offence(s)
+     *         information
+     * @throws HandlerException
+     */
+    public ProsecutionDocument getProsecutionDocument(String resourceUri)
+            throws HandlerException {
+        try {
+            Defendant defendant = prosecutionService.getDefendant(resourceUri);
+            ProsecutionCase prosecutionCase = prosecutionService.getProsecutionCase("/" + defendant.getLinks().get("prosecution-case"));
+            List<Offence> offences = prosecutionService.getOffences(defendant.getLinks().get("offences"));
+            ProsecutionDocument document = new ProsecutionDocument();
+            document.setDefendant(defendant);
+            document.setOffences(offences);
+            document.setProsecutionCase(prosecutionCase);
+            return document;
+        } catch (ProsecutionServiceException pse) {
+            throw new HandlerException("An error occurred when retrieving data from the service: " + pse);
+        }
+    }
+
+    /**
+     * Method to convert an annotated document to a JSON string
+     * 
+     * @param document the annotated document
+     * @param requestId the id of the request being made
+     * @return the mapped JSON string
+     * @throws DocumentInfoCreationException
+     */
+    private String convertToJson(Object document, String requestId)
+            throws DocumentInfoCreationException {
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            return mapper.writeValueAsString(document);
+        } catch (JsonProcessingException e) {
+            throw new DocumentInfoCreationException(
+                    "Could not serialise document info for request id " + requestId);
+        }
+    }
+
+    /**
+     * Method to create the path where the completed document is stored
+     * 
+     * @param type of the prosecution document to store e.g. Ultimatum, SJPn, witness statements
+     * @return the path to store the finished document
+     */
+    private String createLocationLink(ProsecutionType type) {
+        String documentId = String.format("%s-<number>", type.getResource());
+        return String.format("/%s/%s/%s", type.getAssetId(), type.getResource(), documentId);
+    }
+
+    /**
+     * Method to make a DocumentInfoResponse for an Ultimatum
+     * 
+     * @param document The document containing the base information on defendant, prosecution case
+     *        and offences
+     * @param requestId The request id
+     * @return the DocumentInfoResponse for an Ultimatum
+     * @throws HandlerException 
+     */
+    public DocumentInfoResponse getDocumentResponse(ProsecutionDocument document,
```

Comment:
This could be simplified to remove the duplication. The request ID, etc can be included in the log message when the handler exception is caught.
```
public DocumentInfoResponse getDocumentResponse(ProsecutionDocument document,
   String requestId, ProsecutionCaseStatusApi status) throws HandlerException {

   switch(status) {
            case ACCEPTED:
                return getDocumentResponse(document, requestId, status, ProsecutionType.ULTIMATUM);
                break;
            case ULTIMATUM_ISSUED:
                return getDocumentResponse(document, requestId, status, ProsecutionType.SJPN);
                break;
            default:
                throw new HandlerException("Invalid status for Prosecution Case to generate document: " + status);
        }
}

public DocumentInfoResponse getDocumentResponse(ProsecutionDocument data, ProsecutionType type) throws HandlerException {
    DocumentInfoResponse response = new DocumentInfoResponse();
    response.setAssetId(type.getAssetId());
    response.setPath(createLocationLink(type));
    response.setTemplateName(type.getTemplate());
    response.setDescriptionIdentifier(type.getResource());

    try {
            response.setData(convertToJson(data));
        } catch (DocumentInfoCreationException e) {
            throw new HandlerException("Error creating ultimatum document info response for: " + type, e);
     }

     return response;
}
```

Reply:
Okay, will change

Reply:
Done